src/RevCPU.cc

```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
 
 // Initializes a RevThread object.
 // - Moves it to the 'Threads' map
-// - Adds it's ThreadID to the ThreadQueue to be scheduled
-void RevCPU::InitThread(std::shared_ptr<RevThread>& ThreadToInit){
-
+// - Adds it's ThreadID to the ReadyThreads to be scheduled
+void RevCPU::InitThread(std::unique_ptr<RevThread> ThreadToInit){
+  // @Lee, can this be a reference? I'm not looking to transfer ownership of the regfile pointer
```

Comment:
Alright I changed it to &&. Interesting discussion though. Thank you for the insight!

Reply:
In the old code you were constructing a new `unique_ptr` for the function argument, and because `unique_ptr` has no copy constructor, only a move constructor, you had to call the function with `std::move(arg)`. Once in the function, the argument/parameter local variable has ownership of the `unique_ptr`. When you transfer it to its new destination, you have to use `std::move` again.

With `&&` rvalue reference for the function parameter, you are requiring that it only bind to rvalues, so lvalues have to be converted to rvalues by using `std::move` at the call site. Within the function, the parameter name acts as an lvalue to the referent. You have to use `std::move` again to transfer it to its new destination.

The difference between old and new is that a local temporary `unique_ptr` and moves in and out of it were avoided -- now you just use a rvalue reference for the parameter. It still only accepts rvalue arguments, and it still must be converted to a rvalue when it is transferred to somewhere else.